### ISSUE TITLE
App icon notification badge does not appear on Iphones, but works on Android

### ISSUE THREAD

**Issue by Torkilws** [FlutterFlow/flutterflow-issues#1571]
### Has your issue been reported?

- [X] I have searched the existing issues and confirm it has not been reported.
- [X] I give permission for members of the FlutterFlow team to access and test my project for the sole purpose of investigating this issue.

### Current Behavior

I have set up notification for my app. Notification popup-banners works on both Iphone and Android phones.

But the app notification badge (the dot with a number) over the app icon is only present on Android, not iphone.

### Expected Behavior

App icon notification badge is present on both Android and Iphones after a user has received a notification and not opened the app.

### Steps to Reproduce

This one is not so easily reproduceable.

The code needs to be deployed on a Android device or emulator. And for iphone it can only be tested on a real device (as far as I know).

Firebase integration and push notifications also needs to be set up.

### Reproducible from Blank

- [X] The steps to reproduce above start from a blank project.

### Bug Report Code (Required)

ITESlMmA5c9goblH0LnqcMdqhQIgQ0QnU7gnlO1ucx4gJZCsP6QQPs7sV0NKTubmeH9fPmKjmjoz3NLUvuLlCe1dJUuaQaJg0JVPWw3hRkeXRLKXEc2oSnJAJ+xRfUye36azpxYkA7ZfLWQt62GHNq3qNleeY8aSfxBlZ7vfcPo=

### Context

Notifications are sent to users that are invited to events in my app. If an Iphone user miss the notification popup accidentally removes it, they can't see that my app has a notification because of the missing app badge.

### Visual documentation

Using sample images from the web to demonstrate the issue.

### Android:
Android notification popup OK:
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/15776477/a3421aa3-733c-46f5-84b6-03bb3e67080d)

Android notification app icon badge OK:
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/15776477/b8573734-4c22-4d55-885b-a523bac137b3)

### Iphone:
Iphone notification popup OK:
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/15776477/c922d054-ea90-40e5-8874-0eea4670058c)

Iphone notification app icon badge MISSING:
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/15776477/9bf61ebf-9d46-409b-8940-3ab13d0f4cbc)

### Additional Info

![image](https://github.com/FlutterFlow/flutterflow-issues/assets/15776477/dbe18b0a-94c7-4d9b-8d61-7b1029ca9236)

### Environment

```markdown
- FlutterFlow version: v4.0 (released Sept 28. 2023)
- Platform: Windows Web
- Browser name and version: Chrome 117.0.5938.132 (Official Build) (64-bit)
- Operating system and version affected: Windows 11
```

**Comment by hariprasadms** [comment]
Hi @Torkilws - Thanks for submitting the issue. In iOS, app icon badges (notification badges) don't automatically appear on the app icons. Its needs additional code settings to explicitly set the badge count in the app.

**Comment by Torkilws** [comment]
@hariprasadms Ok thanks for confirming this issue. It this something that will be fixed in a flutterflow update, or do I have to add this code manually myself?
If possible, I want to avoid editing the core app code myself so I don't have to create my own develop-branch. But the app icon badges is crucial for my app.

**Comment by hariprasadms** [comment]
Hi @Torkilws - I checked this with dev team and they confirmed that this is currently not readily available in FlutterFlow for iOS devices. It can be achieved in two ways, 
1. Using third party dependency [flutter_app_badger](https://pub.dev/packages/flutter_app_badger) and 
2. Adding something like this to iOS `AppDelegate.{m,swift} `file ( defiantly It needs a bit of reading and understanding). I can confirm at this stage, this should be done at code level. 

I have marked this issue as enhancement/feature request. I would suggest submit a feature requests through the in-app feature request tool. [You can find it in the “Help” menu inside of the FlutterFlow](https://docs.flutterflow.io/getting-started/ui-builder/tool-bar#2.-help). This will help us keep to track of and prioritise feature requests. 

I hope this can help you.

**Comment by wenkaifan0720** [comment]
Hi @Torkilws. I agree that this is something we should be working on. I will add it to our roadmap. Closing this issue for now as it is more of a feature request.

**Comment by Tibop02** [comment]
I have posted this feature request in the Flutterflow community wishlist: 
[https://community.flutterflow.io/general-wishlist/post/app-icon-notification-badge-UfJr7IsjdpcVvqJ](url)

**Comment by ibm777p2** [comment]
I'm waiting for this feature too

**Comment by lsole** [comment]
I also need this feature to be implemented asap, please!

**Comment by Torkilws** [comment]
As this is an essential feature and we have no idea if the FF team will implement this soon or in a year or 2, we just have to keep pestering about it I guess. **It would be amazing with a rough time estimate of when it will be implemented.**

In my case, my (communication) app will not be fully complete without it. I understand it could be built if the code was downloaded and the functionality was built manually. But this kind of destroys the purpose of FF.

**Comment by lsole** [comment]
Looking at the code in flutter_app_badger and assuming it works OK (I haven't tried), I think it shouldn't be so difficult to implement it. Hope the folks at FF put it high on the requests list as this is really a must.

**Comment by ibm777p2** [comment]
> As this is an essential feature and we have no idea if the FF team will implement this soon or in a year or 2, we just have to keep pestering about it I guess. **It would be amazing with a rough time estimate of when it will be implemented.**
> 
> In my case, my (communication) app will not be fully complete without it. I understand it could be built if the code was downloaded and the functionality was built manually. But this kind of destroys the purpose of FF.

I super agree, this is really essential and should been implemented in day one, I don't know why this is not done yet.
Not only your app is not completed without it, many app built with FF are not completed either. I did testing on implementing on Flutter code and it did not work, I tried to put an increment int in the autogenerate code on this function Future _handlePushNotification(RemoteMessage message) async it never worked, looks like it needs to be in the native AppDelegate.swift of iOS to make it work. Make this a priority for once.

**Comment by ibm777p2** [comment]
Still no freaking features in works for the badge count icon in iOS for todays update of new features.
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/5303536/6845ee28-505b-42c5-bc7e-daebf40ba92e)

**Comment by Torkilws** [comment]
What a shame 🙁
Where did you find this?

**Comment by ibm777p2** [comment]
> What a shame 🙁 Where did you find this?

Email of features update of FF

**Comment by wenkaifan0720** [comment]
Hi all. I have bumped this feature request to the priority column. Please be patient while we review and implement this feature. Thank you for your understanding.

**Comment by Wittrup90** [comment]
Upvote

**Comment by ibm777p2** [comment]
This is too basic as iOS feature to be ignored.

**Comment by Torkilws** [comment]
Agreed

**Comment by acx70** [comment]
Needed ASAP

**Comment by ibm777p2** [comment]
Our users are complaining a lot already. I have up to 2000 users and they can't see the badge in the app for iOS.
I'm not sure if flutterflow team understand the magnitud of this feature.

**Comment by acx70** [comment]
I based my app on push notifications but I couldn’t imagine that there wasn’t a badge on the  app icon. There is an entire infrastructure in flutterFlow based on payed features of firebase that supports push notifications but there is not the last piece of the puzzle! It’s like not to have notifications at all!

**Comment by ibm777p2** [comment]
> I based my app on push notifications but I couldn’t imagine that there wasn’t a badge on the app icon. There is an entire infrastructure in flutterFlow based on payed features of firebase that supports push notifications but there is not the last piece of the puzzle! It’s like not having push notifications at all!

I couldn't agree more. Flutterflow team thinks it's only fews out there needing this feature, I'm sure there are more than fews and in each app there could be from 1000 to 10000 users and if you multiply that could be up to millions that actually need this feature. **### Flutterflow team set this for your next delivery** @wenkaifan0720

**Comment by kcs-keshavkushwah** [comment]
Hello, I'm a beginner in flutter.
I'm getting a Firebase notification badge count issue in the Flutter iOS app. My same code is working fine on android app.

**Comment by Torkilws** [comment]
Another awesome update released from the FF team with lots of good stuff. But still no mention of this issue.
An ETA would be very much appreciated.

**Comment by ibm777p2** [comment]
> Another awesome update released from the FF team with lots of good stuff. But still no mention of this issue. An ETA would be very much appreciated.

Yeah, I was gonna say that, this is getting to the point that is ridiculous.

**Comment by Torkilws** [comment]
@wenkaifan0720 how far down on the roadmap is this feature?

**Comment by ibm777p2** [comment]
Can anyone here think about a mobile application that doesn't need a notification badge icon? I would like to know a high efficient mobile app that doesn't need a badge icon.

**Comment by liorzil** [comment]
I would have never imagined in my wildest dreams that this feature is not even available. How is this not a priority?

**Comment by ibm777p2** [comment]
> I would have never imagined in my wildest dreams that this feature is not even available. How is this not a priority?

Yeah this is nuts, just tell me one single high efficient app that doesn't have a badge icon notification. Just imagine uber, airbnb, what's app, signal etc not having this feature, unless flutterflow doesn't want high potential startup to use their platform that's beyond me. @katherineqian

**Comment by Torkilws** [comment]
I also want to add; Not getting notifications while the app is open and in the foreground is also very annoying. Flutterflow notification handling could and should receive some love imo

**Comment by ibm777p2** [comment]
> I also want to add; Not getting notifications while the app is open and in the foreground is also very annoying. Flutterflow notification handling could and should receive some love imo

This is very true, many of my users lost the notifications because of this design, I look at what's app or signal super simple messaging app, signal is even simpler and still getting the notification when the app is open and still taking the user to the destination page pressing the notification while the app is open. The fact that FF apps notification only shows up while the app is running on background or closed and sometimes the notification doesn't take the user to the destination page is a hug flaw on the design, just don't now why it's designed this way. FF is marketed to be the platform to push apps to production the fastest but still not getting the basic functionality right.

**Comment by ibm777p2** [comment]
@wenkaifan0720 what is the progress on this now? is anyone doing the code change for this now?

**Comment by liorzil** [comment]
I'm also waiting urgently for this. I'm nervous that this issue has been opened Oct 2 and we're in January now. When is this expected to be done? Weeks? months? days?

@katherineqian @wenkaifan0720 @hariprasadms

**Comment by wenkaifan0720** [comment]
I understand the importance of this issue. We had some issue with handling this issue if you followed the thread. But this issue will be actively worked on from now on. Sorry for the delay.

**Comment by Torkilws** [comment]
Any news on progress status? Was hoping to include this fix in my next patch.

**Comment by ibm777p2** [comment]
What is the update on this @wenkaifan0720 ? What is the road map?

**Comment by Torkilws** [comment]
Over 4 months and no dot 🥲

**Comment by ibm777p2** [comment]
> Over 4 months and no dot 🥲

@wenkaifan0720 what is the update?

**Comment by acx70** [comment]
That’s what you get when your app depends on someone else code.  I start thinking that the entire flutterFlow business is a big mistake.

**Comment by Torkilws** [comment]
Maybe they are planning to release this fix together with foreground notifications 🤔

**Comment by nhadsall** [comment]
I went to look for instructions to make the badge work, but found this instead. I would have thought this would have been one of the very first features. From a marketing perspective, this is one of the main points of a native app.

**Comment by Torkilws** [comment]
Any update??

**Comment by lsole** [comment]
This will soon be 6 months old... it should be already fixed!

**Comment by ibm777p2** [comment]
@wenkaifan0720 what is the update on this now?

**Comment by KaracsonyMarton** [comment]
same here, really need this feature

**Comment by Torkilws** [comment]
I would really like some more transparency from FF team here.
This feature request for an absolute _essential mobile app feature_ is now soon half a year old. And has supposedly been actively worked on for over 2.5 months. I am checking this issue daily crossing my fingers that there has been any response from the FF team on this thread. This is also the most commented issue of all the FF issues.

Can you at least provide an explanation on why it is taking so long?
- Are you waiting for a dependency?
- Are there still ownership issues involved?
- Are you waiting for another Flutter version?
- Do you not know how to implement it?
- Any other reason?

There are many devs that really needs this feature, many have commented here. And probably more don't event know it yet because they assume it is already in place. Time-sensitive apps are severely held back because of this missing feature. My IOS users often miss important notification because of the missing badge.

@hariprasadms @wenkaifan0720 Could you please provide us with more information? 🙂

**Comment by ibm777p2** [comment]
> I would really like some more transparency from FF team here. This feature request for an absolute _essential mobile app feature_ is now soon half a year old. And has supposedly been actively worked on for over 2.5 months. I am checking this issue daily crossing my fingers that there has been any response from the FF team on this thread. This is also the most commented issue of all the FF issues.
> 
> Can you at least provide an explanation on why it is taking so long?
> 
> * Are you waiting for a dependency?
> * Are there still ownership issues involved?
> * Are you waiting for another Flutter version?
> * Do you not know how to implement it?
> * Any other reason?
> 
> There are many devs that really needs this feature, many have commented here. And probably more don't event know it yet because they assume it is already in place. Time-sensitive apps are severely held back because of this missing feature. My IOS users often miss important notification because of the missing badge.
> 
> @hariprasadms @wenkaifan0720 Could you please provide us with more information? 🙂

My iOS user often miss out urgent messages due to this missed feature.
@hariprasadms @wenkaifan0720 what exactly you need to post us an update here? Are you guys current on vacation due to sprig break? Are you guys currently studying for finals in college ?  Mental break ? Family issues?

**Comment by ibm777p2** [comment]
> > I would really like some more transparency from FF team here. This feature request for an absolute _essential mobile app feature_ is now soon half a year old. And has supposedly been actively worked on for over 2.5 months. I am checking this issue daily crossing my fingers that there has been any response from the FF team on this thread. This is also the most commented issue of all the FF issues.
> > Can you at least provide an explanation on why it is taking so long?
> > 
> > * Are you waiting for a dependency?
> > * Are there still ownership issues involved?
> > * Are you waiting for another Flutter version?
> > * Do you not know how to implement it?
> > * Any other reason?
> > 
> > There are many devs that really needs this feature, many have commented here. And probably more don't event know it yet because they assume it is already in place. Time-sensitive apps are severely held back because of this missing feature. My IOS users often miss important notification because of the missing badge.
> > @hariprasadms @wenkaifan0720 Could you please provide us with more information? 🙂
> 
> My iOS user often miss out urgent messages due to this missed feature. @hariprasadms @wenkaifan0720 what exactly you need to post us an update here? Are you guys current on vacation due to sprig break? Are you guys currently studying for finals in college ? Mental break ? Family issues?

@hariprasadms @wenkaifan0720 We are asking for so much here, we just want 1 or 2 sentences as update per week here.
Please let us know what is preventing you to do that.

**Comment by leighajarett** [comment]
This issue has been fixed and you should see it in the release next week

**Comment by ibm777p2** [comment]
Finally we have that coming
https://community.flutterflow.io/c/whats-new-in-flutterflow/post/what-s-new-in-flutterflow-april-18-2024-xiQAr0TKHHL8oCO
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/5303536/08db6c9c-9f07-43f2-9301-0934387adbd9)

**Comment by whobick** [comment]
Hey all! We just pushed a product update which includes the ability to add badges to your app icon. You can find documentation on how to use the new actions here: https://pub.dev/packages/awesome_notifications

**Important Note:**For Android, badges automatically appear on app icons with push notifications. However, due to [this](https://github.com/firebase/flutterfire/issues/9563) Firebase issue, badges do not automatically appear on iOS app icons at the moment. We'll monitor the situation for a future fix that would allow for automatic badges for IOS.

**Comment by Torkilws** [comment]
@leighajarett is the update out?

**Comment by bdonovan858** [comment]
> Hey all! We just pushed a product update which includes the ability to add badges to your app icon. You can find documentation on how to use the new actions here: https://pub.dev/packages/awesome_notifications
> 
> **Important Note:**For Android, badges automatically appear on app icons with push notifications. However, due to [this](https://github.com/firebase/flutterfire/issues/9563) Firebase issue, badges do not automatically appear on iOS app icons at the moment. We'll monitor the situation for a future fix that would allow for automatic badges for IOS.

Having reviewed the documentation (https://pub.dev/packages/awesome_notifications) for the Awesome Notifications package, it doesn't provide a whole lot of clarity as to how to increment the badge count within a FlutterFlow project (particularly when a push notification is triggered). Could you please provide more detailed guidance on implementing this functionality?

**Comment by bdonovan858** [comment]
Hi @leighajarett, @whobick, and the FF team,

Thanks for your work on the new iOS badge feature. I noticed that the issue was closed, but I wanted to raise a concern that it might not fully address the needs most of us are looking for. The key functionality we're aiming for is for the badge count to increment on the recipient’s device when a push notification is triggered. Having the authenticated user's badge update only after a foreground action is taken on the device doesn't seem as beneficial.

I understand there might be issues with Firebase, but keeping this issue open could be very important. I believe many others share this viewpoint and would appreciate continued efforts to refine this feature.

**Comment by Torkilws** [comment]
@bdonovan858 so the badge functionality is out now?

**Comment by bdonovan858** [comment]
@Torkilws not specifically what you're looking for, but they've added an action that enables you to update your App Badge Count. This doesn't increment the badge for the recipient user on a push notification.
![Screenshot 2024-04-29 at 11 28 11 AM](https://github.com/FlutterFlow/flutterflow-issues/assets/144291520/bb9fd537-6ba8-4445-bd0c-cbbb73731e7f)

**Comment by Torkilws** [comment]
Oh my god really. The whole point was that the badge appears when the user receives a push notification. Doesn't even need to be a number, just a red dot on the icon..

When would you even want to manually update the counter 🤔

**Comment by ibm777p2** [comment]
it's still useless because there is no action that would detect the background notification of the chat messages. This badge update in form of action which mean the end user needs to tap on something in the app in oder to trigger the action then that defeats the purpose of having the badge updated. We want it to be update is because we want the user to tap the app when there is a badge but if this needs the user to tap the app first in order to update the badge defeats the purpose and it's a bit of nonsense.

**Comment by Torkilws** [comment]
@ibm777p2 exactly. We just want it to be the same as on android, where it is working correctly.

@leighajarett this issue is not resolved.

**Comment by ibm777p2** [comment]
@leighajarett I see your firebase department not able to collaborate with your department at the moment in oder to get the ios version working like android because the post of automated badge feature for ios is blocked due to this issue mentioned on the website https://github.com/firebase/flutterfire/issues/9563.
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/5303536/8ce58ae0-456b-4889-824f-db8b0f108d28)

Your both department should prioritize the issues and come up with a timeline when your teams can get this done?

Please consult with firebase, collaborate, prioritize and get this done.
Please find the timeline and post it here.

**Comment by Torkilws** [comment]
@ibm777p2 have you found any solution?

**Comment by ibm777p2** [comment]
> @ibm777p2 have you found any solution?

There are developers manually add code in x-code as a work around solution, the only downside on that is that need to be added every time you publish or test on your device. Not the ideal solution at all.

@leighajarett is there a time line to get this solved https://github.com/firebase/flutterfire/issues/9563. and continue building the actual solution?

**Comment by Torkilws** [comment]
@ibm777p2 The https://github.com/firebase/flutterfire/issues/9563 issue has been closed for a long time. Do you know if there is another issue created for this, to track updates and such?

You would think all flutter users that use firebase notifications would want this fixed.

**Comment by ibm777p2** [comment]
> @ibm777p2 The [firebase/flutterfire#9563](https://github.com/firebase/flutterfire/issues/9563) issue has been closed for a long time. Do you know if there is another issue created for this, to track updates and such?
> 
> You would think all flutter users that use firebase notifications would want this fixed.

@leighajarett the issue that FF mentioned is closed long time ago, will FF continue to develop the badge to be displayed automatically when there is a notification, the way it's now defeat the purpose, the user need to press something in order to execute an action for the badge to show.

Is there a plan to fix this?

**Comment by bdonovan858** [comment]
Yes, that's correct @ibm777p2

**Comment by ibm777p2** [comment]
> Yes, that's correct @ibm777p2

I think your solution works, it's just that changing the cloud function will add the badge for android and ios, then the action block will only clean up the ios device, right? @bdonovan858

**Comment by ibm777p2** [comment]
> Yes, that's correct @ibm777p2

I think this solution works. When you change the cloud function will put the badge in ios and android but when you clear the badge with the action block will only clear for ios, right?

**Comment by Torkilws** [comment]
@bdonovan858 Thank you for coming up with a solution that doesn't involve changing ios code. I will try this, but have a couple of observations:
- You have mentioned **sendUserPushNotificationsTrigger**, but for by case I schedule most of my notifications, so **sendScheduledPushNotifications** will in my case also need to be changed with the badge count.
- To clear the notifications, I assume one would have to use the "Update App Badge Count" (ios only) action and set the count to 0. And that this will clear the badge when opening the app on ios. Is this also needed on Android? I would assume not as I think that the "apns" section in the cloud function is only for ios.
- I also noticed that the Node version used for cloud functions (v16) is deprecated, so this code will probably change before the decomission date (jan 30. 2025).

**Comment by Torkilws** [comment]
I have now changed, deployed and tested this fix. Changing the following functions:
- sendPushNotificationsTrigger
- sendScheduledPushNotifications
- sendUserPushNotificationsTrigger

I received feedback from my friend with iphone, and can confirm that it **worked**!
I also did **not** need to add the action to set the badge count 0 at all, the badge was still removed on opening the app. Thanks again @bdonovan858

**Comment by ibm777p2** [comment]
> I have now changed, deployed and tested this fix. Changing the following functions:
> 
> * sendPushNotificationsTrigger
> * sendScheduledPushNotifications
> * sendUserPushNotificationsTrigger
> 
> I received feedback from my friend with iphone, and can confirm that it **worked**! I also did **not** need to add the action to set the badge count 0 at all, the badge was still removed on opening the app. Thanks again @bdonovan858

@Torkilws did you just put badge: 1 like the image below on these cloud functions and everything work for android and ios?

sendPushNotificationsTrigger
sendScheduledPushNotifications
sendUserPushNotificationsTrigger 
![image](https://github.com/FlutterFlow/flutterflow-issues/assets/5303536/1b554085-794e-42f0-a902-d95d229722dc)

**Comment by Torkilws** [comment]
@ibm777p2 that is right :)

**Comment by ibm777p2** [comment]
> 5\. badge: 1,

How did you guys made work? I change the cloud functions and still not working for me. I still don't see the red badge.

**Comment by ibm777p2** [comment]
> > 5. badge: 1,
> 
> Finally made work too. Thanks @Torkilws @bdonovan858

I think I still need to use the action block to clean it, the red badge doesn't go away by looking at the message.

**Comment by Torkilws** [comment]
@ibm777p2 I see now that it might have worked "out of the box" for me because I already had this custom action added to `main.dart`:

```
import 'package:clear_all_notifications/clear_all_notifications.dart';

Future clearAllNotifications() async {
  // Add your function code here!
  if (!isWeb) {
    await ClearAllNotifications.clear();
  }
}
```

Using
`clear_all_notifications: ^1.0.0`

**Comment by ibm777p2** [comment]
Super crazy the community actually solved this and the developers of flutterflow did not contributed much for our success.
This solutions is not perfect but it's everything you expected.

**Comment by azizirac** [comment]
Hi @leighajarett, @whobick, and the FF team,

I tried to add `badge: 1` to my functions and the push notification works. I have updated my entire logic to store the number of notifications of a users in firebase and retrieve it when sending a new notification. 

Based on https://developer.apple.com/documentation/usernotifications/generating-a-remote-notification, the `badge` variable is a correct solution. Can't Flutterflow team add it to their functions and allow users to set the counter when sending a notification with Send Notification Action? This will allow everyone to leverage it without making changes on the FF functions on their own.

**Comment by AdrKacz** [comment]
@azizirac @Torkilws @ibm777p2 I've been playing with notifications on iOS lately, here is a solution you can adapt to have a dynamic badge.

The idea remains to update the Cloud Function, I provide at the end of the message some guidance for the @FlutterFlowEng team to implement a none-invasive quick fix. Happy to discuss more about it.

Instead of setting the value to `1`, you want measure what should be the value of the badge for each users that receive a notification. In the example I share, I want the badge value to be the number of unread thread, but it can really be whatever you want.

For this I:
1. created a new field in my `group` object *(the thread)* that I will use to store who didn't read the last message send
2. updated the app so that when the user open the thread, it update the `group` object and remove the its `id` from the list of user who didn't read the latest messages
3. updated the Cloud Function to count in how many `groups` the user `id` appears.

```js
async function sendPushNotifications(snapshot) {
  const notificationData = snapshot.data();
  const title = notificationData.notification_title || "";
  const body = notificationData.notification_text || "";
  const imageUrl = notificationData.notification_image_url || "";
  const sound = notificationData.notification_sound || "";
  const parameterData = notificationData.parameter_data || "";
  const targetAudience = notificationData.target_audience || "";
  const initialPageName = notificationData.initial_page_name || "";
  const userRefsStr = notificationData.user_refs || "";
  const batchIndex = notificationData.batch_index || 0;
  const numBatches = notificationData.num_batches || 0;
  const status = notificationData.status || "";

  if (status !== "" && status !== "started") {
    console.log(`Already processed ${snapshot.ref.path}. Skipping...`);
    return;
  }

  if (title === "" || body === "") {
    await snapshot.ref.update({ status: "failed" });
    return;
  }

  const userRefs = userRefsStr === "" ? [] : [...new Set(userRefsStr.trim().split(","))]; // Deduplicate user references
  const badgeMap = new Map(); // Maps badge values to sets of tokens

  if (userRefs.length > 0) {
    for (const userRef of userRefs) {
      const userDoc = firestore.doc(userRef);
      const unreadGroupsCountSnapshot = await firestore
        .collection("groups")
        .where("last_message_unread_by", "array-contains", userDoc)
        .count()
        .get();

      const badgeValue = unreadGroupsCountSnapshot.data().count;
      if (!badgeMap.has(badgeValue)) {
        badgeMap.set(badgeValue, new Set());
      }

      const userTokens = await userDoc
        .collection(kFcmTokensCollection)
        .get();

      userTokens.docs.forEach((tokenDoc) => {
        const tokenData = tokenDoc.data();
        if (typeof tokenData.fcm_token !== "undefined") {
          badgeMap.get(badgeValue).add(tokenData.fcm_token);
        }
      });
    }
  } else {
    // Assign all tokens a badge value of 0 (default when no userRefs provided)
    badgeMap.set(0, new Set());
    var userTokensQuery = firestore.collectionGroup(kFcmTokensCollection);
    // Handle batched push notifications by splitting tokens up by document
    // id.
    if (numBatches > 0) {
      userTokensQuery = userTokensQuery
        .orderBy(admin.firestore.FieldPath.documentId())
        .startAt(getDocIdBound(batchIndex, numBatches))
        .endBefore(getDocIdBound(batchIndex + 1, numBatches));
    }
    const userTokens = await userTokensQuery.get();
    userTokens.docs.forEach((token) => {
      const data = token.data();
      const audienceMatches =
        targetAudience === "All" || data.device_type === targetAudience;
      if (audienceMatches && typeof data.fcm_token !== undefined) {
        badgeMap.get(0).add(data.fcm_token);
      }
    });
  }

  const messageBatches = [];
  for (const [badgeValue, tokensSet] of badgeMap.entries()) {
    const tokensArr = Array.from(tokensSet);
    for (let i = 0; i < tokensArr.length; i += 500) {
      const tokensBatch = tokensArr.slice(i, Math.min(i + 500, tokensArr.length));
      const messages = {
        notification: {
          title,
          body,
          ...(imageUrl && { imageUrl }),
        },
        data: {
          initialPageName,
          parameterData,
        },
        android: {
          notification: {
            ...(sound && { sound }),
          },
        },
        apns: {
          payload: {
            aps: {
              badge: badgeValue,
              ...(sound && { sound }),
            },
          },
        },
        tokens: tokensBatch,
      };
      messageBatches.push(messages);
    }
  }

  var numSent = 0;
  await Promise.all(
    messageBatches.map(async (messages) => {
      const response = await admin.messaging().sendEachForMulticast(messages);
      numSent += response.successCount;
    })
  );

  await snapshot.ref.update({ status: "succeeded", num_sent: numSent });
}
```

Now, to @FlutterFlowEng. You will agree that it is not fare to expect your users to update a Cloud Function for such a common use case, especially as your target audience is *"not too much"* technical *(don't get me wrong, it is a great tool even for technical folks, and by making things more accessible you attracted a lot of people who couldn't build an app otherwise)*.

To resolve this issue with a simple fix you can implement in one or two hours, I'd suggest:
1. Update the notifications code widget to take as parameter an optional `badgeValue` *(it is important that it is optional for backward compatibility)*
2. Let the user define any logic to provide the value of the badge (in my case for example, I would use your custom code builder to count the number of `group` with the field `last_message_unread_by` containing the user id, and such for all users I'm sending the notification to, a different person with a different project will use a different logic of course)
3. Update the default Cloud Function you provide as follow:

```js

async function sendPushNotifications(snapshot) {
  // ...

  const badgeValue = notificationData.badgeValue || 0;

  // ...

  const tokensArr = Array.from(tokens);
  var messageBatches = [];
  for (let i = 0; i < tokensArr.length; i += 500) {
    const tokensBatch = tokensArr.slice(i, Math.min(i + 500, tokensArr.length));
    const messages = {
      notification: {
        title,
        body,
        ...(imageUrl && { imageUrl: imageUrl }),
      },
      data: {
        initialPageName,
        parameterData
      },
      android: {
        notification: {
          ...(sound && { sound: sound }),
        },
      },
      apns: {
        payload: {
          aps: {
              ...(badgeValue && { badge: badgeValue }), // This will ensure that not sending any badge won't break anything
              ...(sound && { sound }),
           },
        },
      },
      tokens: tokensBatch,
    };
    messageBatches.push(messages);
  }

  // ... END OF THE FUNCTION
}
```

I understand this is not the perfect solution, especially regarding (2) as you could imagine a better design that will automatically account for multiple users so there is no need for inner loops *(as notifications will often be send to more than one users)*. However, it is a **working** solution, as as of now you have nothing, so it will be better anyways. I would advise starting with this fix you can implement and push to production with the next release, while working on a more robust and long term solution if it's one of your priorities.

Thank you,
